core/src/ids.rs

```diff
@@ -18,270 +18,186 @@
 
 //! Base identifiers for messaging primitives.
 
-use blake2_rfc::blake2b;
-use core::convert::TryInto;
-use scale_info::{
-    scale::{Decode, Encode},
-    TypeInfo,
-};
+use blake2::{digest::typenum::U32, Blake2b, Digest};
+pub use gprimitives::{ActorId, CodeId, MessageId, ReservationId};
 
-/// Hash length used in gear protocol.
-pub const HASH_LENGTH: usize = 32;
-
-/// Hash type used in gear protocol.
-pub type Hash = [u8; HASH_LENGTH];
+/// BLAKE2b-256 hasher state.
+type Blake2b256 = Blake2b<U32>;
 
 /// Creates a unique identifier by passing given argument to blake2b hash-function.
 ///
 /// # SAFETY: DO NOT ADJUST HASH FUNCTION, BECAUSE MESSAGE ID IS SENSITIVE FOR IT.
-pub fn hash(argument: &[u8]) -> Hash {
-    let blake2b_hash = blake2b::blake2b(HASH_LENGTH, &[], argument);
+pub fn hash(data: &[u8]) -> [u8; 32] {
+    let mut ctx = Blake2b256::new();
+    ctx.update(data);
+    ctx.finalize().into()
+}
 
-    blake2b_hash
-        .as_bytes()
-        .try_into()
-        .expect("we set hash len; qed")
+/// Creates a unique identifier by passing given argument to blake2b hash-function.
+///
+/// # SAFETY: DO NOT ADJUST HASH FUNCTION, BECAUSE MESSAGE ID IS SENSITIVE FOR IT.
+pub fn hash_of_array<T: AsRef<[u8]>, const N: usize>(array: [T; N]) -> [u8; 32] {
+    let mut ctx = Blake2b256::new();
+    for data in array {
+        ctx.update(data);
+    }
+    ctx.finalize().into()
 }
 
-/// Declares data type for storing any kind of id for gear-core,
-/// which stores 32 bytes under the hood.
-#[macro_export]
-macro_rules! declare_id {
-    ($name:ident: $doc: literal) => {
-        #[doc=$doc]
-        #[derive(
-            Clone,
-            Copy,
-            Default,
-            Eq,
-            Hash,
-            Ord,
-            PartialEq,
-            PartialOrd,
-            Decode,
-            Encode,
-            parity_scale_codec::MaxEncodedLen,
-            derive_more::From,
-            TypeInfo,
-        )]
-        pub struct $name($crate::ids::Hash);
-
-        impl $name {
-            /// Creates new id.
-            ///
-            /// Never use it in production!
-            pub const fn test_new(hash: $crate::ids::Hash) -> Self {
-                Self(hash)
-            }
-
-            /// Returns id as bytes array.
-            pub fn into_bytes(self) -> $crate::ids::Hash {
-                self.0
-            }
-        }
+/// Program identifier.
+pub type ProgramId = ActorId;
 
-        impl From<$name> for $crate::ids::Hash {
-            fn from(val: $name) -> $crate::ids::Hash {
-                val.0
-            }
-        }
+pub mod prelude {
+    //! The purpose of this module is to make it easier to import `gprimitives` extensions.
```

Comment:
Usually `prelude` is used in examples. Also, any IDE provides auto-import feature. So I'm not sure this module really should exist

Reply:
Well, this seems more convenient than `use gear_core::ids::{ActorId, ActorIdExt, CodeId, CodeIdExt}` (also `cargo fmt` will sort alphabetically...)

Reply:
im ok having there prelude, but as I've said before to @StackOverflowExcept1on, for me importing in caller code `use gear_core::ids::*` looks ok due to module's basic utility, so exts can be implemented 1 module upper